source/common/upstream/outlier_detection_impl.cc

```diff
@@ -259,15 +299,44 @@ DetectorConfig::DetectorConfig(const envoy::config::cluster::v3::OutlierDetectio
       max_ejection_time_jitter_ms_(static_cast<uint64_t>(PROTOBUF_GET_MS_OR_DEFAULT(
           config, max_ejection_time_jitter, DEFAULT_MAX_EJECTION_TIME_JITTER_MS))),
       successful_active_health_check_uneject_host_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(
-          config, successful_active_health_check_uneject_host, true)) {}
+          config, successful_active_health_check_uneject_host, true)),
+      validation_visitor_(validation_visitor) {
+
+  // Store extensions' config. It will be used to create extensions monitors
+  // when host are added to the cluster.
+  extensions_config_ = config.monitors();
```

Comment:
You may need to get the monitor factory and load the monitor config here. Then, only do the creation of Monitor instances when we try to create the host detection monitor to ensure there is no any exception (like no factory, invalid config, etc.) will be throwed at that time.

Reply:
For example:
```
// Factory should load the configuration and return a lambda.
using MonitorFn = std::function<MonitorPtr()>;

class MonitorFactory : public Envoy::Config::TypedFactory {
public:
  ~MonitorFactory() override = default;

  virtual MonitorFn createMonitor(const std::string& name, const Protobuf::Message& config,
                                   MonitorFactoryContext& context) PURE;

  std::string category() const override { return "envoy.outlier_detection_monitors"; }
};

```

Reply:
That was good idea! Thanks for suggesting it. So, when config is received factory for each extension only validates the config and creates a callback which creates an extension. When callback is called, the config is guaranteed to be valid.